### PR TITLE
k8s/query+ds: catch query errors, fix datasourceProvider and expose prometheus

### PR DESCRIPTION
**Pull request by ryantxu**
This extracts some minor fixes from https://github.com/grafana/grafana/pull/84969 and https://github.com/grafana/grafana-enterprise/pull/6366

### CODE REVIEW

File: pkg/registry/apis/datasource/register.go [hunk 1]
```diff
@@ -69,7 +69,8 @@ func RegisterAPIService(
 	all := pluginStore.Plugins(context.Background(), plugins.TypeDataSource)
 	ids := []string{
 		"grafana-testdata-datasource",
-		//	"prometheus",
+		"prometheus",
+		"graphite",
```
Comment (ryantxu) [inline pkg/registry/apis/datasource/register.go:73]: Exposing in embedded mode makes testing query service delegation easier (I have been commenting out the condition check)

File: pkg/registry/apis/query/query.go [hunk 2]
```diff
@@ -135,6 +135,17 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
...
+	if rsp == nil && err != nil {
+		rsp = &backend.QueryDataResponse{Responses: make(backend.Responses)}
+		for _, q := range req.Request.Queries {
+			rsp.Responses[q.RefID] = backend.DataResponse{
+				Status: backend.Status(code),
+				Error:  err,
+			}
+		}
+	}
 	return rsp, err
```
Comment (toddtreece) [inline pkg/registry/apis/query/query.go:149]: should the error also be set on the response in all cases?

Comment (ryantxu) [reply pkg/registry/apis/query/query.go:149]: the error is already handled upstream -- the missing part was the status code getting lost and just returning an empty response (and error that did not point to any query)

Comment (toddtreece) [reply pkg/registry/apis/query/query.go:149]: what about in situations like line 128? is it intentional that the validation error is overwriting `err` and returned on line 149?

**Review by toddtreece — APPROVED** [review]

File: .vscode/launch.json [hunk 3]
```diff
@@ -23,6 +23,20 @@
           "--secure-port=8443", 
           "--runtime-config=testdata.datasource.grafana.app/v0alpha1=true"]
     },
+    {
+      "name": "Run API Server (query-localhost)",
```
Comment (andresmgot) [inline .vscode/launch.json:27]: I have a bunch of configurations like these locally. I wonder if we should `gitignore` this file and add some tooling to generate them.

**Review by andresmgot — APPROVED** [review]

File: pkg/registry/apis/query/query.go (hunk 2 above)
Comment (ryantxu) [reply pkg/registry/apis/query/query.go:149]: It is intentional -- though I am not sure it is a good idea yet.

File: pkg/registry/apis/datasource/plugincontext.go [hunk 4]
```diff
@@ -39,29 +43,42 @@ type PluginContextWrapper interface {
...
+	contextProvider *plugincontext.Provider) ScopedPluginDatasourceProvider {
+	return &cachingDatasourceProvider{
 		dsService:       dsService,
 		dsCache:         dsCache,
 		contextProvider: contextProvider,
 	}
 }
 
-type defaultPluginDatasourceProvider struct {
+type cachingDatasourceProvider struct {
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/registry/apis/datasource/plugincontext.go:54]: dooh -- already merged -- but called it caching becasue it requires the CacheService 🤷🏻‍♂️

File: .vscode/launch.json (hunk 3 above)
Comment (ryantxu) [reply .vscode/launch.json:27]: Good question -- I think the value of simple debugging from vscode is worth keeping it out of gitignore.  When you say you have a bunch... should we add them?

Comment (andresmgot) [reply .vscode/launch.json:27]: I have one for cloudinfo and then one for testdata and prometheus. We'd end up with one per core source, I don't know if that's too much noise or if there is a better solution for that

**Pull request merged**